### PR TITLE
os/tools : Introduce a binary header generator 'mkheader.py'

### PR DESCRIPTION
**Pull request by YiHyunJin**
Python script generates header for Binary manager.

header information :

 / header size (2byte) / binary type (2byte) / binary size (4byte)
 / Binary name (16byte) / Binary version (12byte)
 / Kernel version (8byte) / Jump address (4byte)

### CODE REVIEW

**Comment by seinfra** [comment]
Target : [9a4e229a91533cd389eb9bfad846ae2f61da73b6] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [9a4e229a91533cd389eb9bfad846ae2f61da73b6] - Code Rule Check OK.

**Comment by jsdosa** [comment]
I think there should be 
more explanation of what is used in the code. 
Currently, only the function description is provided.

File: os/tools/mkheader.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+#
+# parameter information
+# argv[1] is filepath
+# argv[2] is file type
+# argv[3] is Kernel version
+# argv[4] is Binary name
+# argv[5] is Binary version
+############################################################################
+
+header_size = 48
```
Comment (jeongchanKim) [inline os/tools/mkheader.py]: Can we calculate this size with type like sizeof in C?

File: os/tools/mkheader.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+    fp.seek(binary_ver_point, 0)
+    fp.write(binary_ver)
+
+    fp.seek(kernel_ver_point, 0)
+    fp.write(kernel_ver)
+    
+    fp.seek(jump_addr_point, 0)
+    
+    # parsing _vector_start address from elf information.
```
Comment (jeongchanKim) [inline os/tools/mkheader.py]: _vector_start is only for ARM architecture. Please add comments the details.

File: os/tools/mkheader.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+############################################################################
+
+header_size = 48
+file_path  = sys.argv[1]
+binary_type = sys.argv[2]
+kernel_ver = sys.argv[3]
+binary_name = sys.argv[4]
+binary_ver = sys.argv[5]
+
+binary_ver_point = 24
```
Comment (jeongarmy) [inline os/tools/mkheader.py]: I think it is better to define size values of each variable, not calculated offset.

File: os/tools/mkheader.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    file_size = fp.tell()
+    fp.close()
+    fp = open(file_path, 'wb')
+
+    # header initialization
+    for i in range(0, header_size):
+        fp.write('\000')
+
+    fp.seek(0,0)
+    hex_dize = "%04x" %header_size
```
Comment (jeongarmy) [inline os/tools/mkheader.py]: please fix typo

File: os/tools/mkheader.py (hunk 3 above)
Comment (jeongarmy) [reply os/tools/mkheader.py]: some codes would be changed, for example, in line 69
fp.write(binary_name);
offset += binary_name_size;

fp.seek(offset, 0);
fp.write(binary_ver);

Comment (YiHyunJin) [reply os/tools/mkheader.py]: I modified it.

File: os/tools/mkheader.py (hunk 2 above)
Comment (YiHyunJin) [reply os/tools/mkheader.py]: I added.

File: os/tools/mkheader.py (hunk 4 above)
Comment (YiHyunJin) [reply os/tools/mkheader.py]: I fixed it.

**Comment by seinfra** [comment]
Target : [ef4deb5460b120ea70d4facb2cb2dd351ce69ab8] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [ef4deb5460b120ea70d4facb2cb2dd351ce69ab8] - Code Rule Check (C++) OK.

File: os/tools/mkheader.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    fp.write(kernel_ver)
+
+    OFFSET += KERNEL_VERSION_S
+    fp.seek(OFFSET,0)
+    fp.seek(jump_addr_point, 0)
+    
+    # parsing _vector_start address from elf information.
+    # _vector_start is only for ARM architecture. so it
+    # operates in the ARM architecture.
+    os.system('readelf -s ../build/output/bin/tinyara | grep _vector_start > addr_file')
```
Comment (jeongchanKim) [inline os/tools/mkheader.py]: `../build/output/bin/tinyara` can be changed.
So please define a variable in the top.

**Comment by seinfra** [comment]
Target : [c52c3bed3efd0ff2b44b5ada9594b35d60ee266b] -

**Comment by seinfra** [comment]
Target : [c52c3bed3efd0ff2b44b5ada9594b35d60ee266b] -

**Comment by seinfra** [comment]
Target : [35813bac9817cc5cedaac42eba4a7aa613edf500] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [35813bac9817cc5cedaac42eba4a7aa613edf500] - Code Rule Check (C++) OK.

File: os/tools/mkheader.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+    for i in range(0, header_size):
+        fp.write('\000')
+
+    OFFSET = START
+
+    fp.seek(OFFSET,0)
+    fp.write(struct.pack('h', header_size))
+    OFFSET += HEADER_SIZE_S
+
+    if binary_type == 'kernel' or binary_type == 'KERNEL' :
```
Comment (jeongchanKim) [inline os/tools/mkheader.py]: I think 'kernel' and 'elf' is not in the same hierarchy.
how about 'bin' and 'elf'?

**Comment by seinfra** [comment]
Target : [1f9ea7c592b5b93ba298485c82295688a2bf5d47] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [1f9ea7c592b5b93ba298485c82295688a2bf5d47] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [6bdba8e6949e4611587cc02967f31e393bd08704] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [6bdba8e6949e4611587cc02967f31e393bd08704] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [0ebc13f1b86c875e4c8fd45c8b1d5e522d5db263] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [0ebc13f1b86c875e4c8fd45c8b1d5e522d5db263] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [d3f0932363fc4aad73e945846f0ec5fdea54f33c] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [d3f0932363fc4aad73e945846f0ec5fdea54f33c] - Code Rule Check (C++) OK.

**Pull request merged**